### ISSUE TITLE
(JavaScript error) RangeError: Maximum call stack size exceeded. 

### ISSUE THREAD

**Issue by almightynassar** [wkhtmltopdf/wkhtmltopdf#3262]
(Confirmed issue in 0.12.3. Earlier versions seem to have no issue. Have not tested on 0.12.4 as no repo binaries for my main OS)

Using Docker with Alpine base. Configured a Laravel environment using barryvdh and knplabs wrappers for wkhtmltopdf. I found issues with generating PDFs that require javascript (for fancy graphs).

Running wkhtmltopdf on the temporary file generated by the wrapper scripts (inside the Alpine Docker container):

```
/data/storage # wkhtmltopdf --version
wkhtmltopdf 0.12.3
/data/storage # wkhtmltopdf --debug-javascript knp_snappy585d2a423d0de6.09687853.html test.pdf
Loading page (1/2)
Warning: undefined:0 RangeError: Maximum call stack size exceeded.
Warning: undefined:0 RangeError: Maximum call stack size exceeded.
Printing pages (2/2)                                               
Done
```

Running wkhtmltopdf on the same temporary file, this time on my main OS:

```
/data/storage$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 16.04.1 LTS
Release:	16.04
Codename:	xenial
/data/storage$ wkhtmltopdf --version
wkhtmltopdf 0.12.2.4
/data/storage$ wkhtmltopdf --debug-javascript knp_snappy585d2a423d0de6.09687853.html test1.pdf
Loading page (1/2)
Printing pages (2/2)                                               
Done
```

As you can see, same file but 0.12.3 reaches the call stack limit.

Short of manually compiling either 0.12.4 or an earlier release, is there a way to increase that call stack limit?

Thank you for your time.

**Comment by almightynassar** [comment]
Also, my other machine has 0.12.1 in it's repository list. I had installed that version and the above file processes correctly. It seems the issue is limited to the 0.12.3 release and maybe 0.12.4 once I have a workable installation ready for testing.

**Comment by ashkulz** [comment]
@almightynassar: looks like you're using the distro-provided version i.e. without the patched Qt. Can you try the static linux binaries (they might work as-is)?

**Comment by almightynassar** [comment]
Thank you so much for your help over the holiday season. I have encountered a new issue that may be due to Docker itself, but I'll provide you with the details to see what insights you can give me.

I ran the following inside of the Docker container:

```
/tmp # wget http://download.gna.org/wkhtmltopdf/0.12/0.12.4/wkhtmltox-0.12.4_linux-generic-amd64.tar.xz
/tmp # unxz wkhtmltox-0.12.4_linux-generic-amd64.tar.xz
/tmp # tar -xvf wkhtmltox-0.12.4_linux-generic-amd64.tar 
/tmp # cd wkhtmltox/bin
/tmp/wkhtmltox/bin # ./wkhtmltopdf 
sh: ./wkhtmltopdf: not found
/tmp # /tmp/wkhtmltox/bin/wkhtmltopdf --version
sh: /tmp/wkhtmltox/bin/wkhtmltopdf: not found
```

I moved the whole untarred directory to a volume directly accessible to the host and ran the following on local machine:

```
[almightynassar@localhost data]$ ./wkhtmltox/bin/wkhtmltopdf --version
wkhtmltopdf 0.12.4 (with patched qt)
```

So the static binary works on my host machine but does not work on the docker machine. I'm used to BASH so it might be a shell thing; any insight is appreciated.

**Comment by ashkulz** [comment]
In the docker container, try `ls -al`. It may be that `/tmp` is mounted noexec i.e you cannot execute binaries from it, which is a security measure.

**Comment by almightynassar** [comment]
```
/tmp/wkhtmltox/bin # ls -al
total 77564
drwxr-xr-x    2 root     root          4096 Jan  2 21:53 .
drwxr-xr-x    6 root     root          4096 Jan  2 21:53 ..
-rwxr-xr-x    1 root     root      39666344 Jan 20  2016 wkhtmltoimage
-rwxr-xr-x    1 root     root      39745960 Jan 20  2016 wkhtmltopdf
```

Decided to check the mount point.

```
/tmp/wkhtmltox/bin # mount | grep /tmp
/dev/sda1 on /tmp type ext4 (rw,relatime,errors=remount-ro,data=ordered)
```

Checked the error message too.

```
/tmp/wkhtmltox/bin # /tmp/wkhtmltox/bin/wkhtmltopdf --version
sh: /tmp/wkhtmltox/bin/wkhtmltopdf: not found
/tmp/wkhtmltox/bin # echo $?
127
```

Is there a nice handy list of required shared lib files that the binary is looking for that may be on ubuntu (host) but not alpine (container)? I want to make sure I have them installed on my container.

Just as a sanity check, copied the folder to /usr/local/bin where another script I use is stored (composer). Ran both, wkhtmltopdf still failed but composer still ran so I am pretty sure it isn't a noexec flag.

```
/tmp/wkhtmltox/bin # cp -R /tmp/wkhtmltox /usr/local/bin/
/tmp/wkhtmltox/bin # ls -al /usr/local/bin/
total 1696
drwxr-xr-x    3 root     root          4096 Jan  2 22:17 .
drwxr-xr-x   12 root     root          4096 Jan  2 21:57 ..
-rwxr-xr-x    1 root     root       1710687 Dec 22 03:24 composer
-rwxrwxr-x    1 root     root          1119 Dec 18 15:35 env.sh
-rwxrwxr-x    1 root     root           130 Dec 18 23:20 wkhtmltoimage.sh
-rwxrwxr-x    1 root     root           128 Dec 18 23:18 wkhtmltopdf.sh
drwxr-xr-x    6 root     root          4096 Jan  2 22:17 wkhtmltox
/tmp/wkhtmltox/bin # cd /usr/local/bin/wkhtmltox/bin/
/usr/local/bin/wkhtmltox/bin # ./wkhtmltopdf 
sh: ./wkhtmltopdf: not found
/usr/local/bin/wkhtmltox/bin # echo $?
127
/usr/local/bin/wkhtmltox/bin # /usr/local/bin/composer -V
Composer version 1.2.4 2016-12-06 22:00:51
```

**Comment by almightynassar** [comment]
Some more digging. Ran ldd on the binary:

```
/tmp/wkhtmltox/bin # ldd ./wkhtmltopdf 
	/lib64/ld-linux-x86-64.so.2 (0x558144809000)
	libXrender.so.1 => /usr/lib/libXrender.so.1 (0x7fa97dc1a000)
	libfontconfig.so.1 => /usr/lib/libfontconfig.so.1 (0x7fa97d9df000)
	libfreetype.so.6 => /usr/lib/libfreetype.so.6 (0x7fa97d730000)
	libXext.so.6 => /usr/lib/libXext.so.6 (0x7fa97d520000)
	libX11.so.6 => /usr/lib/libX11.so.6 (0x7fa97d1fd000)
	libz.so.1 => /lib/libz.so.1 (0x7fa97cfe7000)
	libdl.so.2 => /lib64/ld-linux-x86-64.so.2 (0x558144809000)
	librt.so.1 => /lib64/ld-linux-x86-64.so.2 (0x558144809000)
	libpthread.so.0 => /lib64/ld-linux-x86-64.so.2 (0x558144809000)
	libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x7fa97cc97000)
	libm.so.6 => /lib64/ld-linux-x86-64.so.2 (0x558144809000)
	libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x7fa97ca84000)
	libc.so.6 => /lib64/ld-linux-x86-64.so.2 (0x558144809000)
Error loading shared library ld-linux-x86-64.so.2: No such file or directory (needed by ./wkhtmltopdf)
	libexpat.so.1 => /usr/lib/libexpat.so.1 (0x7fa97c864000)
	libpng16.so.16 => /usr/lib/libpng16.so.16 (0x7fa97c637000)
	libxcb.so.1 => /usr/lib/libxcb.so.1 (0x7fa97c417000)
	libXau.so.6 => /usr/lib/libXau.so.6 (0x7fa97c214000)
	libXdmcp.so.6 => /usr/lib/libXdmcp.so.6 (0x7fa97c00e000)
```

Did a search on the [package repository](https://pkgs.alpinelinux.org/contents?file=ld-linux-*&path=&name=&branch=&repo=main&arch=) for the file and found that it is provided by the **libc6-compat** library

```
/tmp/wkhtmltox/bin # apk add libc6-compat --no-cache
/tmp/wkhtmltox/bin # ldd ./wkhtmltopdf 
	/lib64/ld-linux-x86-64.so.2 (0x55d4d5afe000)
	libXrender.so.1 => /usr/lib/libXrender.so.1 (0x7fda9ca7a000)
	libfontconfig.so.1 => /usr/lib/libfontconfig.so.1 (0x7fda9c83f000)
	libfreetype.so.6 => /usr/lib/libfreetype.so.6 (0x7fda9c590000)
	libXext.so.6 => /usr/lib/libXext.so.6 (0x7fda9c380000)
	libX11.so.6 => /usr/lib/libX11.so.6 (0x7fda9c05d000)
	libz.so.1 => /lib/libz.so.1 (0x7fda9be47000)
	libdl.so.2 => /lib64/ld-linux-x86-64.so.2 (0x55d4d5afe000)
	librt.so.1 => /lib64/ld-linux-x86-64.so.2 (0x55d4d5afe000)
	libpthread.so.0 => /lib64/ld-linux-x86-64.so.2 (0x55d4d5afe000)
	libstdc++.so.6 => /usr/lib/libstdc++.so.6 (0x7fda9baf7000)
	libm.so.6 => /lib64/ld-linux-x86-64.so.2 (0x55d4d5afe000)
	libgcc_s.so.1 => /usr/lib/libgcc_s.so.1 (0x7fda9b8e4000)
	libc.so.6 => /lib64/ld-linux-x86-64.so.2 (0x55d4d5afe000)
	ld-linux-x86-64.so.2 => /lib/ld-linux-x86-64.so.2 (0x7fda9b658000)
	libexpat.so.1 => /usr/lib/libexpat.so.1 (0x7fda9b438000)
	libpng16.so.16 => /usr/lib/libpng16.so.16 (0x7fda9b20b000)
	libxcb.so.1 => /usr/lib/libxcb.so.1 (0x7fda9afeb000)
	libXau.so.6 => /usr/lib/libXau.so.6 (0x7fda9ade8000)
	libXdmcp.so.6 => /usr/lib/libXdmcp.so.6 (0x7fda9abe2000)
/tmp/wkhtmltox/bin # /tmp/wkhtmltox/bin/wkhtmltopdf --version
sh: /tmp/wkhtmltox/bin/wkhtmltopdf: not found
/tmp/wkhtmltox/bin # echo $?
127
```

Still failing.... did a quick sanity check that the container is running x64....

```
/tmp/wkhtmltox/bin # uname -a
Linux f47c576f86bb 4.4.0-57-generic #78-Ubuntu SMP Fri Dec 9 23:50:32 UTC 2016 x86_64 Linux
```

**Comment by ashkulz** [comment]
Uhm, you're still trying in `/tmp`. And when you copied it to `/usr/local`, you got `wkhtmltopdf.sh` which is not what the static binary has. Looks like an installation issue to me...

**Comment by almightynassar** [comment]
**TL;DR** wkhtmltopdf is built for glibc, and Alpine runs musl libc. This difference is big enough to cause wkhtmltopdf to not run properly under Alpine. Installing a hacky glibc compatibility package in Alpine gets me part of the way there (just need to find a few shared libraries), but since the official line from Alpine is to not support glibc in the hopes that developers build against musl means that I'll always have to tinker away at the container. I have chosen to use debian:jessie instead which works perfectly fine for my needs (it just isn't as tiny as I would like).

Thanks again for your help. If you ever release a musl build of wkhtmltopdf then I will gladly test in a Docker container.
-----------------------------------------------------------------------------------------------------------------------------

wkhtmltopdf.sh is a wrapper script that just calls xvfb and the Alpine repo version of wkhtmltopdf. That system works fine but cannot render the javascript properly because it hits the "Maximum call stack size exceeded" error. I kept the repo version installed to ensure that the necessary libraries were installed (and then tried it all again with it uninstalled).

I should have made it more clear that running an executable in /tmp was not a problem.

```
/tmp # echo '#!/bin/sh' > example.sh
/tmp # echo "echo 'Hello'" >> example.sh
/tmp # chmod +x example.sh 
/tmp # /tmp/example.sh 
Hello
/tmp # cp /usr/local/bin/composer ./
/tmp # /tmp/composer -V
Composer version 1.3.0 2016-12-24 00:47:03
/tmp # ls -al
total 1780
drwxrwxrwx    2 root     root            38 Jan  3 05:59 .
drwxr-xr-x   19 root     root          4096 Jan  3 05:57 ..
-rwxr-xr-x    1 root     root       1811606 Jan  3 05:58 composer
-rwxr-xr-x    1 root     root            23 Jan  3 05:57 example.sh
```

My sneaking suspicion is that Alpine does not have the shared library or feature that wkhtmltopdf is looking for. Probably because it uses  musl libc instead of glibc.

Installing and running with glibc isntalled gives a different error message:

```
/tmp # apk --no-cache add ca-certificates openssl
/tmp # wget -q -O /etc/apk/keys/sgerrand.rsa.pub https://raw.githubusercontent.com/sgerrand/alpine-pkg-glibc/master/sgerrand.rsa.pub
/tmp # wget https://github.com/sgerrand/alpine-pkg-glibc/releases/download/2.23-r3/glibc-2.23-r3.apk
/tmp # apk add glibc-2.23-r3.apk
/tmp # /tmp/wkhtmltox/bin/wkhtmltopdf --version
/tmp/wkhtmltox/bin/wkhtmltopdf: error while loading shared libraries: libXrender.so.1: cannot open shared object file: No such file or directory
/tmp # /usr/bin/wkhtmltopdf --version # Default Alpine installation for comparison
wkhtmltopdf 0.12.3
/tmp # cp /usr/bin/wkhtmltopdf ./ # Copying default Alpine installation to /tmp to prove it isn't problem
/tmp # /tmp/wkhtmltopdf --version
wkhtmltopdf 0.12.3
```

Alpine does not set $LD_LIBRARY_PATH, and we need to set the library lookup to look for glibc libraries first:

```
/tmp # export LD_LIBRARY_PATH=/usr/glibc-compat/lib:/usr/lib:/lib64:/lib
/tmp # /tmp/wkhtmltox/bin/wkhtmltopdf --version
/tmp/wkhtmltox/bin/wkhtmltopdf: /usr/lib/libstdc++.so.6: no version information available (required by /tmp/wkhtmltox/bin/wkhtmltopdf)
/tmp/wkhtmltox/bin/wkhtmltopdf: /usr/lib/libstdc++.so.6: no version information available (required by /tmp/wkhtmltox/bin/wkhtmltopdf)
Segmentation fault (core dumped)
```

So all I need now is a version of libstdc++.so compiled for glibc and I should be able to run wkhtmltopdf. But that is too difficult to maintain compared to what I can do with debian, so I switched back over.

As a proof of concept, in the Debian container:

```
root@a4ba7e020505:/tmp# apt-get install wget xz-utils
root@a4ba7e020505:/tmp# wget http://download.gna.org/wkhtmltopdf/0.12/0.12.4/wkhtmltox-0.12.4_linux-generic-amd64.tar.xz
root@a4ba7e020505:/tmp# unxz wkhtmltox-0.12.4_linux-generic-amd64.tar.xz 
root@a4ba7e020505:/tmp# tar -xf wkhtmltox-0.12.4_linux-generic-amd64.tar
root@a4ba7e020505:/tmp# /tmp/wkhtmltox/bin/wkhtmltopdf --version
wkhtmltopdf 0.12.4 (with patched qt)
```

Updating my wrapper script to use the /tmp binary instead of the debian repo version works fine. There is still some variance across different development containers but I believe that is separate to this particular issue.
